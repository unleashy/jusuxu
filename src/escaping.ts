export function escapeTextForHtml(text: string): string {
  return text.replace(/[<&]/g, match => {
    if (match === "<") {
      return "&#x3c;";
    } else {
      // match === "&"
      return "&#x26;";
    }
  });
}
