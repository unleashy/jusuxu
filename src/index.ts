export type { Component } from "./element";
export * from "./jsx-runtime";
export * from "./renderer";
