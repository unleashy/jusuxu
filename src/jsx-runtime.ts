export function jsx() {}

export const jsxs = jsx;
