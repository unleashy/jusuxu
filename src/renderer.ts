export class Renderer {
  render(element: JSX.Element): string {
    return `<!DOCTYPE html>${this.renderFragment(element)}`;
  }

  renderFragment(element: JSX.Element): string {
    return element.render();
  }
}
