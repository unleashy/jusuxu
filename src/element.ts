import { escapeTextForHtml } from "./escaping";

export interface Element {
  render: () => string;
}

export type Props = Record<string, unknown> & { children?: unknown };

export class HtmlElement implements Element {
  constructor(readonly tagName: string, readonly props: Props) {}

  render(): string {
    const { children, ...props } = this.props;
    if (VOID_ELEMENTS.includes(this.tagName)) {
      return `<${this.tagName}${this.renderAttributes(props)}>`;
    } else {
      // prettier-ignore
      return `<${this.tagName}${this.renderAttributes(props)}>${this.renderChildren(children)}</${this.tagName}>`;
    }
  }

  private renderAttributes(props: Omit<Props, "children">): string {
    const processedAttrs = Object.entries(props)
      .map(([k, v]) => {
        if (v === true) {
          return k;
        } else if (v === false || typeof v === "undefined" || v === null) {
          return false;
        } else {
          return `${k}="${v}"`;
        }
      })
      .filter(it => it); // remove false boolean attrs

    if (processedAttrs.length === 0) {
      return "";
    } else {
      return " " + processedAttrs.join(" ");
    }
  }

  private renderChildren(children: unknown): string {
    switch (typeof children) {
      case "string":
        return escapeTextForHtml(children);

      case "object":
        if (Array.isArray(children)) {
          return children.map(child => this.renderChildren(child)).join("");
        } else {
          return (children as Element | null)?.render() ?? "";
        }

      default:
        return "";
    }
  }
}

const VOID_ELEMENTS = [
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
];
