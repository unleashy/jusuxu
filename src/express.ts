import { Renderer } from "./renderer";
import { jsx } from "./jsx-runtime";

type ExpressEngine = (
  path: string,
  options: Record<string, unknown>,
  callback: (e: Error | null, rendered?: string) => void
) => void;

export function createEngine(): ExpressEngine {
  const renderer = new Renderer();

  return (path, props, callback) => {
    const mod = require(path);
    const Component = mod.__esModule === true ? mod.default : mod;
    if (typeof Component !== "function") {
      throw new TypeError(
        `View must default-export a component function, got ${typeof Component} instead`
      );
    }

    const rendered = renderer.render(jsx(Component, props));
    callback(null, rendered);
  };
}
