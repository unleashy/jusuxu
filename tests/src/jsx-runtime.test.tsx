import { Fragment, jsx, jsxs } from "../../src";
import {
  ComponentElement,
  FragmentElement,
  HtmlElement
} from "../../src/element";

describe(jsx, () => {
  it("constructs Elements properly", () => {
    expect(jsx("a", { href: "link" })).toBeInstanceOf(HtmlElement);
    expect(jsx(jest.fn(), { foo: "bar" })).toBeInstanceOf(ComponentElement);
    expect(jsx(Fragment, { children: [] })).toBeInstanceOf(FragmentElement);
  });

  it.each([
    "",
    " ",
    "\f",
    "\n",
    "\t",
    "\0",
    "/",
    ">",
    "!",
    "?",
    "a ",
    "b\f",
    "c\n",
    "d\t",
    "F/",
    "G>",
    "h\0"
  ])("throws on tag name %j", tag => {
    expect(() => jsx(tag, {})).toThrowErrorMatchingSnapshot();
  });

  it.each(["A!", "B?", "cð"])("does not throw on tag name %j", tag => {
    expect(() => jsx(tag, {})).not.toThrow();
  });

  it.each([
    "",
    " ",
    "\f",
    "\n",
    "\t",
    "\0",
    "/",
    ">",
    "<",
    "=",
    "'",
    '"',
    " a",
    "b ",
    "c/",
    "d>",
    "e<",
    "F=",
    "G'",
    'H"'
  ])("throws on attribute name %j", attr => {
    expect(() => jsx("div", { [attr]: "" })).toThrowErrorMatchingSnapshot();
  });

  it.each(["a[]", "b:c", "dę", "f^g"])(
    "does not throw on attribute name %j",
    attr => {
      expect(() => jsx("div", { [attr]: "" })).not.toThrow();
    }
  );
});

describe("jsxs", () => {
  it("is the exact same as jsx", () => {
    expect(jsxs).toBe(jsx);
  });
});

describe("jsx syntax", () => {
  it("does the same thing as calling `jsx`", () => {
    expect(<a href="link">child</a>).toEqual(
      jsx("a", { href: "link", children: "child" })
    );
  });
});
