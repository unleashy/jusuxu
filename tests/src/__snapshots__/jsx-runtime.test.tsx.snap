// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx throws on attribute name " " 1`] = `"Attribute name \\" \\" is not valid."`;

exports[`jsx throws on attribute name " a" 1`] = `"Attribute name \\" a\\" is not valid."`;

exports[`jsx throws on attribute name "" 1`] = `"Attribute name \\"\\" is not valid."`;

exports[`jsx throws on attribute name "'" 1`] = `"Attribute name \\"'\\" is not valid."`;

exports[`jsx throws on attribute name "/" 1`] = `"Attribute name \\"/\\" is not valid."`;

exports[`jsx throws on attribute name "<" 1`] = `"Attribute name \\"<\\" is not valid."`;

exports[`jsx throws on attribute name "=" 1`] = `"Attribute name \\"=\\" is not valid."`;

exports[`jsx throws on attribute name ">" 1`] = `"Attribute name \\">\\" is not valid."`;

exports[`jsx throws on attribute name "\\"" 1`] = `"Attribute name \\"\\\\\\"\\" is not valid."`;

exports[`jsx throws on attribute name "\\f" 1`] = `"Attribute name \\"\\\\f\\" is not valid."`;

exports[`jsx throws on attribute name "\\n" 1`] = `"Attribute name \\"\\\\n\\" is not valid."`;

exports[`jsx throws on attribute name "\\t" 1`] = `"Attribute name \\"\\\\t\\" is not valid."`;

exports[`jsx throws on attribute name "\\u0000" 1`] = `"Attribute name \\"\\\\u0000\\" is not valid."`;

exports[`jsx throws on attribute name "F=" 1`] = `"Attribute name \\"F=\\" is not valid."`;

exports[`jsx throws on attribute name "G'" 1`] = `"Attribute name \\"G'\\" is not valid."`;

exports[`jsx throws on attribute name "H\\"" 1`] = `"Attribute name \\"H\\\\\\"\\" is not valid."`;

exports[`jsx throws on attribute name "b " 1`] = `"Attribute name \\"b \\" is not valid."`;

exports[`jsx throws on attribute name "c/" 1`] = `"Attribute name \\"c/\\" is not valid."`;

exports[`jsx throws on attribute name "d>" 1`] = `"Attribute name \\"d>\\" is not valid."`;

exports[`jsx throws on attribute name "e<" 1`] = `"Attribute name \\"e<\\" is not valid."`;

exports[`jsx throws on tag name " " 1`] = `"Tag name \\" \\" is not valid."`;

exports[`jsx throws on tag name "!" 1`] = `"Tag name \\"!\\" is not valid."`;

exports[`jsx throws on tag name "" 1`] = `"Tag name \\"\\" is not valid."`;

exports[`jsx throws on tag name "/" 1`] = `"Tag name \\"/\\" is not valid."`;

exports[`jsx throws on tag name ">" 1`] = `"Tag name \\">\\" is not valid."`;

exports[`jsx throws on tag name "?" 1`] = `"Tag name \\"?\\" is not valid."`;

exports[`jsx throws on tag name "\\f" 1`] = `"Tag name \\"\\\\f\\" is not valid."`;

exports[`jsx throws on tag name "\\n" 1`] = `"Tag name \\"\\\\n\\" is not valid."`;

exports[`jsx throws on tag name "\\t" 1`] = `"Tag name \\"\\\\t\\" is not valid."`;

exports[`jsx throws on tag name "\\u0000" 1`] = `"Tag name \\"\\\\u0000\\" is not valid."`;

exports[`jsx throws on tag name "\\u0000" 2`] = `"Tag name \\"\\\\u0000\\" is not valid."`;

exports[`jsx throws on tag name "F/" 1`] = `"Tag name \\"F/\\" is not valid."`;

exports[`jsx throws on tag name "G>" 1`] = `"Tag name \\"G>\\" is not valid."`;

exports[`jsx throws on tag name "a " 1`] = `"Tag name \\"a \\" is not valid."`;

exports[`jsx throws on tag name "b\\f" 1`] = `"Tag name \\"b\\\\f\\" is not valid."`;

exports[`jsx throws on tag name "c\\n" 1`] = `"Tag name \\"c\\\\n\\" is not valid."`;

exports[`jsx throws on tag name "d\\t" 1`] = `"Tag name \\"d\\\\t\\" is not valid."`;

exports[`jsx throws on tag name "h\\u0000" 1`] = `"Tag name \\"h\\\\u0000\\" is not valid."`;

exports[`jsx throws on tag name '	' 1`] = `"Tag name \\"\\\\t\\" is not valid."`;

exports[`jsx throws on tag name '
' 1`] = `"Tag name \\"\\\\n\\" is not valid."`;

exports[`jsx throws on tag name '' 1`] = `"Tag name \\"\\\\f\\" is not valid."`;

exports[`jsx throws on tag name ' ' 1`] = `"Tag name \\" \\" is not valid."`;

exports[`jsx throws on tag name '' 1`] = `"Tag name \\"\\" is not valid."`;
